warehouse/ingest-core/src/main/java/datawave/ingest/mapreduce/handler/edge/ProtobufEdgeDataTypeHandler.java

```diff
@@ -741,6 +609,183 @@ public long process(KEYIN key, RawRecordContainer event, Multimap<String,Normali
+        
+        Multimap<String,NormalizedContentInterface> mSource = depthFirstList.get(edgeDef.getSourceFieldName());
+        Multimap<String,NormalizedContentInterface> mSink = depthFirstList.get(edgeDef.getSinkFieldName());
+        
+        // get the activity date from the event fields map
+        long activityDate = getActivityDate();
+        boolean validActivityDate = validateActivityDate(activityDate, event.getDate());
+        boolean activityEqualsEvent = false;
+        
+        // If the activity date is valid check to see if it is on the same day as the event date
+        if (validActivityDate) {
+            activityEqualsEvent = compareActivityAndEvent(activityDate, event.getDate());
+        }
+        
+        EdgeKey.DATE_TYPE date_type = getEdgeKeyDateType(validActivityDate, activityEqualsEvent, event.getDate());
+        
+        for (String sourceSubGroup : sourceSubGroups) {
+            for (NormalizedContentInterface ifaceSource : mSource.get(sourceSubGroup)) {
+                for (String sinkSubGroup : sinkSubGroups) {
+                    for (NormalizedContentInterface ifaceSink : mSink.get(sinkSubGroup)) {
+                        EdgeDataBundle edgeValue = createEdge(edgeDef, event, ifaceSource, sourceGroup, sourceSubGroup, ifaceSink, sinkGroup, sinkSubGroup);
+                        if (edgeValue != null) {
+                            
+                            // have to write out the keys as the edge values are generated, so counters get updated
+                            // and the system doesn't timeout.
+                            edgesCreated += writeEdgesForDateType(edgeValue, context, contextWriter, date_type);
+                            
+                            if (this.enableMetadata) {
+                                registerEventMetadata(edgeValue, edgeDef);
+                                
+                            }
+                        }
+                        
+                    }
+                }
+            }
+            
+        }
+        return edgesCreated;
+    }
+    
+    private void normalizeAndGroupFields(Multimap<String,NormalizedContentInterface> fields) {
+        Multimap<String,NormalizedContentInterface> tmp = null;
+        for (Entry<String,NormalizedContentInterface> e : fields.entries()) {
+            NormalizedContentInterface value = e.getValue();
+            String subGroup = null;
+            if (value instanceof GroupedNormalizedContentInterface) {
+                subGroup = ((GroupedNormalizedContentInterface) value).getSubGroup();
+            }
+            String fieldName = getGroupedFieldName(value);
+            tmp = depthFirstList.get(fieldName);
+            if (tmp == null) {
+                tmp = HashMultimap.create();
+            }
+            tmp.put(subGroup, value);
+            depthFirstList.put(fieldName, tmp);
+            
+            normalizedFields.put(fieldName, value);
+        }
+    }
+    
```

Comment:
I like the refactor because it improves readability of the code.  Given the amount of refactoring, it seems prudent to ensure there are unit tests that provide sufficient regression testing - to guarantee that exactly the same number and kind of edges are created as before.  I don't know that the existing tests provide that type of testing.  I recommend adding "black box" edge tests for this purpose.

Along these lines, I recommend separating the changes to functionality from the refactor changes.  It seems that the vast majority of the changes are for refactoring.

Reply:
Thanks for adding these tests!